### ISSUE TITLE
Want to use oxc_linter as crate

### ISSUE THREAD

**Issue by ityuany** [oxc-project/oxc#7509]
I really like the oxc series of tools.

We have a scenario where oxc_linter is intended to be used directly as rust crate. Could you please send a package to support it

**Comment by Boshen** [comment]
That's extended maintenance burden for us. Can you use `git = "https://github.com/oxc-project/oxc.git"` first and send PRs to make it accessible before we release it crates.io?

**Comment by ityuany** [comment]
@Boshen  This allows me to download napi, once additionally

<img width="1337" alt="image" src="https://github.com/user-attachments/assets/b506e9c1-bc9f-49d4-87d2-432ed3e28c84">

**Comment by ityuany** [comment]
<img width="998" alt="image" src="https://github.com/user-attachments/assets/70ec15d1-4e0e-43dc-aec5-5f139a416d86">

Does this seem to have new requirements for my rust version?

**Comment by ityuany** [comment]
@Boshen 

Sorry, I'm a rust novice and used to work more on node and the front end. I don't quite understand why releasing oxc to Crate.io would add an extra maintenance burden, my understanding he should just be more than a publish action, perhaps I exist in the knowledge of the blind, can answer it. In addition, it may be important to me that OXC be released to Crate.io so I can use it directly, and I'd still like to support it if possible. Thank you very much

**Comment by nrayburn-tech** [comment]
Releasing as a crate means that it now needs to be maintained and supported for clients to use. This includes things like APIs that make sense for clients (structs, methods), changelog, etc. This maintenance is the extra burden. The action of just publishing it is unlikely to be an issue.

**Comment by camc314** [comment]
> <img alt="image" width="1030" src="https://private-user-images.githubusercontent.com/13724214/390618374-38b18e99-6b32-4440-83f6-88b8127f979a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI5Nzg3MDksIm5iZiI6MTczMjk3ODQwOSwicGF0aCI6Ii8xMzcyNDIxNC8zOTA2MTgzNzQtMzhiMThlOTktNmIzMi00NDQwLTgzZjYtODhiODEyN2Y5NzlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTMwVDE0NTMyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwODQzOGQzYWMzNGZmMWU1N2Q5NzFhYmExYTk5MjdjYTA3ZjRmYzEyZDc2N2NjNWNlYTRhYmQ0NDk2NDk5MWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5E2Icc8dBkhOMx3ac0fhmOO3xaEtuPt7jmK4qW4VBzk">
> Does this seem to have new requirements for my rust version?

Looks like this was stablized in 1.81.0 . i would try updating your rust version to solve the problem

https://github.com/rust-lang/rust/pull/120924

**Comment by ityuany** [comment]
> Releasing as a crate means that it now needs to be maintained and supported for clients to use. This includes things like APIs that make sense for clients (structs, methods), changelog, etc. This maintenance is the extra burden. The action of just publishing it is unlikely to be an issue.

I understand, I think what you said makes a lot of sense, thank you for your response. Maybe I should try to understand the overall code design of oxc more deeply, and then try to take on this part of the work actively?

**Comment by nrayburn-tech** [comment]
I believe that is why Boshen was suggesting. My understanding of it is below, although you should confirm with him.

Use the library with the GitHub URL for now. As you find issues, submit PRs to make improvements. Once things have been improved in a way that it is better to consume as a library and Boshen is comfortable with it not adding a significant maintenance burden, publishing as a crate becomes a much more realistic option (publishing as a crate isn’t guaranteed at this point, but it also isn’t completely off the table).

**Comment by ityuany** [comment]
<img width="552" alt="image" src="https://github.com/user-attachments/assets/a1675fb8-e41a-4243-9e76-9c7a988ca936">

**Comment by Boshen** [comment]
> I tried to install it using the git address and found the private module
> 
> <img alt="image" width="552" src="https://private-user-images.githubusercontent.com/13724214/394676582-a1675fb8-e41a-4243-9e76-9c7a988ca936.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5MTE1OTAsIm5iZiI6MTczMzkxMTI5MCwicGF0aCI6Ii8xMzcyNDIxNC8zOTQ2NzY1ODItYTE2NzVmYjgtZTQxYS00MjQzLTllNzYtOWM3YTk4OGNhOTM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjExVDEwMDEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhMjEwYmRjNmJjMzU5MDQxNmNkMWFiYTJkYmQ2ODIxOGE5OWQ3OTJjYmM4NDJkZDI0YWE4MmU5YWU5N2QxYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IK9Ct7r7NX04kIW2s4loRpEEx0__0oegRCQw4ah5eSg"> <img alt="image" width="879" src="https://private-user-images.githubusercontent.com/13724214/394676996-bede4061-5aea-4f2e-a7a1-e6ab10ab65f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5MTE1OTAsIm5iZiI6MTczMzkxMTI5MCwicGF0aCI6Ii8xMzcyNDIxNC8zOTQ2NzY5OTYtYmVkZTQwNjEtNWFlYS00ZjJlLWE3YTEtZTZhYjEwYWI2NWYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjExVDEwMDEzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmNzE3YTRhMzI1NGE2N2UwMGMwYzI0YWNhNTQyNmJjM2QwMjk4YjVhM2E5Njk1ZjczYjUwM2I0ZDA1ZmIyMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.brNRPtEOZydCG6F7E8QflndIDPXZSQi_1a96VsMRddk">

Can you PR?

**Comment by ityuany** [comment]
I think maybe I found the right way to use it, but there seems to be some compatibility issues

<img width="907" alt="image" src="https://github.com/user-attachments/assets/be9efedd-2499-4f41-a48c-9ceaab06eab2">


```rust
use std::{path::Path, rc::Rc, sync::Arc};

use oxc_allocator::Allocator;
use oxc_linter::{
  AllowWarnDeny, FixKind, LintFilter, LintFilterKind, LinterBuilder,
  RuleCategory,
};
use oxc_parser::Parser;
use oxc_semantic::SemanticBuilder;
use oxc_span::SourceType;

fn main() -> std::io::Result<()> {
  let path = Path::new("hello.js");
  let source_text = std::fs::read_to_string(path)?;

  let allocator = Allocator::default();
  let source_type = SourceType::from_path(path).unwrap();
  let ret = Parser::new(&allocator, &source_text, source_type).parse();

  let semantic_builder = SemanticBuilder::new()
    .with_cfg(true)
    .with_scope_tree_child_ids(true);

  let semantic_ret = semantic_builder.build(&ret.program);
  let semantic = Rc::new(semantic_ret.semantic);
  let module_record = Arc::new(oxc_linter::ModuleRecord::new(
    path,
    &ret.module_record,
    semantic,
  ));

  let linter = LinterBuilder::default()
    .with_filters(vec![
      LintFilter::new(AllowWarnDeny::Warn, "no-debugger").unwrap(),
      LintFilter::new(AllowWarnDeny::Deny, "no-console").unwrap(),
      LintFilter::new(AllowWarnDeny::Warn, "no-commonjs").unwrap(),
      LintFilter::new(AllowWarnDeny::Warn, RuleCategory::Correctness).unwrap(),
    ])
    .with_fix(FixKind::None)
    .build();

  let diagnostics = linter.run(path, semantic, module_record);

  for message in diagnostics {
    println!("{:?}", message.error);
  }

  Ok(())
}

```

<img width="1723" alt="image" src="https://github.com/user-attachments/assets/aa1995c5-53e4-4a6e-bbdb-a1b8564650f4">

**Comment by ityuany** [comment]
![image](https://github.com/user-attachments/assets/c4b41ec8-e167-452a-aabf-bfc96bd612ad)

**Comment by Boshen** [comment]
@ityuany use 0.39.0 for now, there's a problem with 0.40.1.

And this is why I don't want to release it as a crate, this process is already burdening.

**Comment by ityuany** [comment]
<img width="1065" alt="image" src="https://github.com/user-attachments/assets/362c8150-1e7a-45be-ae69-4ca92df409c0" />

I've found that this works fine, but it creates a confusion for my team members that all `tags` use `oxlint_v0.15.0`, which makes the versions of `oxc_ast`, `oxc_span`, ... This will make the versions of `oxc_ast`, `oxc_span`, ... less intuitive and clear. But at least this way he is working. I'm going to use it this way for now, and wait for the next dear Boshen to solve the conflict problem better!
